### ISSUE TITLE
TURN "Test this server" always results in "Error: No working ICE candidates returned by the TURN server"

### ISSUE THREAD

**Issue by prismopensource** [nextcloud/spreed#5563]
## Description
On the settings/admin/talk page at the TURN section, the "Test this server" action always fails on Safari with Mac OS 11. When using latest Firefox without changing the turnserver:port and secret fields, everything works as advertised. Makes no difference testing an already configured TURN server or directly after adding a new one.

## Steps to reproduce
1. Add a TURN server by adding turnserver:port and secret
2. Click the "Test this server" button

### Expected behaviour
A succes message with "Ok: Successful ICE candidates returned by the TURN server"

### Actual behaviour
An error message with "Error: No working ICE candidates returned by the TURN server"

## Talk app

**Talk app version:** 10.0.7

**Custom Signaling server configured:** no

**Custom TURN server configured:** yes

**Custom STUN server configured:** no

## Server configuration detail

**Operating system:** Linux 4.19.0-16-cloud-amd64 #1 SMP Debian 4.19.181-1 (2021-03-19) x86_64

**Webserver:** Apache/2 (fpm-fcgi)

**Database:** mysql 10.5.9

**PHP version:** 7.4.18
Modules loaded: Core, date, libxml, openssl, pcre, sqlite3, zlib, bcmath, calendar, ctype, curl, dom, hash, fileinfo, filter, ftp, gd, gettext, SPL, iconv, intl, json, mbstring, session, pcntl, PDO, standard, pdo_sqlite, Phar, posix, Reflection, mysqlnd, SimpleXML, soap, sockets, sodium, pdo_mysql, exif, tokenizer, xml, xmlreader, xmlwriter, xsl, zip, mysqli, cgi-fcgi, gmp, imagick, imap, redis, Zend OPcache

**Nextcloud version:** 20.0.9 - 20.0.9.1

**Updated from an older Nextcloud/ownCloud or fresh install:** fresh

**Where did you install Nextcloud from:** unknown

<details><summary>Signing status</summary>

Array
(
)

</details>

<details><summary>List of activated apps</summary>

```
Enabled:
 - accessibility: 1.6.0
 - activity: 2.13.4
 - bruteforcesettings: 2.1.0
 - calendar: 2.2.1
 - cloud_federation_api: 1.3.0
 - comments: 1.10.0
 - contacts: 3.5.1
 - contactsinteraction: 1.1.0
 - dashboard: 7.0.0
 - dav: 1.16.2
 - federatedfilesharing: 1.10.2
 - federation: 1.10.1
 - files: 1.15.0
 - files_pdfviewer: 2.0.1
 - files_rightclick: 0.17.0
 - files_sharing: 1.12.2
 - files_trashbin: 1.10.1
 - files_versions: 1.13.0
 - files_videoplayer: 1.9.0
 - firstrunwizard: 2.9.0
 - groupfolders: 8.2.1
 - issuetemplate: 0.7.0
 - logreader: 2.5.0
 - lookup_server_connector: 1.8.0
 - mail: 1.9.5
 - nextcloud_announcements: 1.9.0
 - notifications: 2.8.0
 - oauth2: 1.8.0
 - password_policy: 1.10.1
 - photos: 1.2.3
 - privacy: 1.4.0
 - provisioning_api: 1.10.0
 - recommendations: 0.8.0
 - serverinfo: 1.10.0
 - settings: 1.2.0
 - sharebymail: 1.10.0
 - spreed: 10.0.7
 - support: 1.3.0
 - survey_client: 1.8.0
 - systemtags: 1.10.0
 - talk_simple_poll: 1.3.0
 - text: 3.1.0
 - theming: 1.11.0
 - twofactor_backupcodes: 1.9.0
 - twofactor_totp: 5.0.0
 - updatenotification: 1.10.0
 - user_external: 1.0.0
 - user_status: 1.0.1
 - viewer: 1.4.0
 - weather_status: 1.0.0
 - workflowengine: 2.2.0
Disabled:
 - admin_audit
 - encryption
 - files_external
 - user_ldap

```
</details>

<details><summary>Configuration (config/config.php)</summary>

```
{
    "instanceid": "***REMOVED SENSITIVE VALUE***",
    "passwordsalt": "***REMOVED SENSITIVE VALUE***",
    "secret": "***REMOVED SENSITIVE VALUE***",
    "trusted_domains": [
        "***REMOVED SENSITIVE VALUE***"
    ],
    "datadirectory": "***REMOVED SENSITIVE VALUE***",
    "dbtype": "mysql",
    "version": "20.0.9.1",
    "overwrite.cli.url": "https:\/\/cloud.***REMOVED SENSITIVE VALUE***",
    "dbname": "***REMOVED SENSITIVE VALUE***",
    "dbhost": "***REMOVED SENSITIVE VALUE***",
    "dbport": "",
    "dbtableprefix": "oc_",
    "mysql.utf8mb4": true,
    "dbuser": "***REMOVED SENSITIVE VALUE***",
    "dbpassword": "***REMOVED SENSITIVE VALUE***",
    "installed": true,
    "user_backends": [
        {
            "class": "OC_User_IMAP",
            "arguments": [
                "mail.***REMOVED SENSITIVE VALUE***",
                993,
                "ssl",
                "***REMOVED SENSITIVE VALUE***",
                true,
                true
            ]
        }
    ],
    "mail_smtpmode": "smtp",
    "mail_smtpauthtype": "LOGIN",
    "mail_sendmailmode": "smtp",
    "mail_smtpauth": 1,
    "updater.release.channel": "stable",
    "memcache.distributed": "\\OC\\Memcache\\Redis",
    "redis": {
        "host": "***REMOVED SENSITIVE VALUE***",
        "port": 6379
    },
    "memcache.locking": "\\OC\\Memcache\\Redis",
    "memcache.local": "\\OC\\Memcache\\Redis",
    "maintenance": false,
    "debug": true
}
```
</details>

**Are you using external storage, if yes which one:** local

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** IMAP

## Client configuration

**Browser:** Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Safari/605.1.15 without any content blockers active

**Operating system:** macOS 11.2.3

## Logs

<details><summary>Web server error log</summary>

```
empty 
```
</details>

<details><summary>Nextcloud log</summary>

```
nothing related to this bug
```
</details>

<details><summary>Browser log</summary>

[Info] Creating PeerConnection with – Object (admin-settings.js, line 380)
Object
[Info] Received candidates – Array (0) (admin-settings.js, line 380)
Array (0)
> Selected Element
< <a data-v-f2e0410a class="icon icon-error" data-original-title="null" style></a>
[Info] Creating PeerConnection with – {iceServers: Array, iceTransportPolicy: "relay"} (admin-settings.js, line 380)
{iceServers: Array, iceTransportPolicy: "relay"}Object
[Info] Received candidates – [] (0) (admin-settings.js, line 380)
</details>

**Comment by nickvergessen** [comment]
Same with Safari on iPad

**Comment by mogluk** [comment]
Same issue here...Nextcloud version 21.0.1 on ubuntu 20.04.2, turn and stun server configured with tls...everything works fine on talk 11.1.2 (independent on client side). With talk 11.2.1 I do get the error "No working ICE candidates returned by the TURN server" too.

**Comment by acsfer** [comment]
Just want to mention this is NOT tied to Mac/Safari. On Windows/Chrome it can be reproduced.
But on the console, i think servers are correctly fetched despite the error.

![image](https://user-images.githubusercontent.com/12234510/118678086-00070100-b7fd-11eb-8e14-5876599a9860.png)

**Comment by szaimen** [comment]
I could just reproduce this with Firefox: Although Turn is configured for UDP and TCP connections, the test in Firefox currently only works for UDP. The TCP test returns the same issue.

**Comment by danielepercivaldi** [comment]
Hi. I can confirm the comment by @mogluk and @acsfer . @szaimen, I open port 5349/UDP but the test still fails.
In version 10.07 for NC 20 it works and there are not parameters about turn (turn only, turns only, turn and turns). I cannot understand their meaning.

**Comment by hoatienii** [comment]
same error “Error: No working ICE candidates returned by the TURN server”. don't know how to fix

**Comment by prismopensource** [comment]
> same error “Error: No working ICE candidates returned by the TURN server”. don't know how to fix

The workaround at the moment is to use Firefox.

**Comment by hoatienii** [comment]
> > same error “Error: No working ICE candidates returned by the TURN server”. don't know how to fix
> 
> The workaround at the moment is to use Firefox.

error on firefox too. i can not call, only chat.

**Comment by nickvergessen** [comment]
This topic here is only about the test functionality. A patch has been backported to fix it. Calling functionality itself is not impacted by the bug.

**Comment by nickvergessen** [comment]
Should be fixed with https://github.com/nextcloud/spreed/pull/5665

**Comment by velocity08** [comment]
Hi All

we are seeing the same issue, can someone please share what logs we need to review in the singling server relation to ICE ?

would be greatly appreciated.

ta

**Comment by nickvergessen** [comment]
Update to the latest version

**Comment by velocity08** [comment]
I have and still getting the same issue.
wondering if my symptoms are the same but the root cause is different.

any help with the log location will help with my troubleshooting?

im not seeing anything defined in /etc/signaling/server.conf for logs.

any assistance in troubleshooting would be greatly appreciated :)

""Cheers
G

**Comment by velocity08** [comment]
scratch that last request i have fixed it :)

its now showing as working in FF and other browsers.

thanks

""G

**Comment by HeyWatchOutDude** [comment]
I cant get a working ICE candidate from my TURN server.

**Comment by DominiqueFuchs** [comment]
Same here. Encountered this bug for the first time now, so I can't say anything about recent version history behavior – but right now I'm running Talk 11.2.2 on Server 21.0.2 and:

- In Safari, TURN server test fails, not receiving any candidates
  -  can actually be verified by console output: candidate array is empty
- Same instance, admin panel in Chrome: Test works, candidates are received as expected.

@nickvergessen Any chance this regressed? Could this issue be reopened, so that it's on the list for re-confirmation?

**Comment by HeyWatchOutDude** [comment]
Fixed it .... had a permission issue.
(Copy of the ssl certificate / privkey wasn't owned by the turnserver user)

Note:
chown -R turnserver:turnserver /etc/coturn/
chmod -R 755 /etc/coturn/
chmod 644 /etc/coturn/certs/$FqdnTurnServer.fullchain.pem
chmod 600 /etc/coturn/certs/$FqdnTurnServer.privkey.pem

Source: https://blog.wydler.eu/2020/04/04/eigener-turn-server-fuer-jitsi-meet-bereitstellen

ICE candidate: OK!

**Comment by mauzilla** [comment]
Latest version NC (21.0.3) and Talk (11.2.2) - Does not work in Chrome, however Firefox works.

**Comment by andyxheli** [comment]
Running NC 21.0.3 with Talk App V 11.3.1 and works with no issues when i click on test server check comes back using Chrome.

**Comment by bbx-github** [comment]
I have the same issue:
- nextcloud 23.0.0.10, talk 13.0.3
- video connection is working
- talk test error message in firefox 96 and chromium 98: "Error: No working ICE candidates returned by the TURN server"

**Comment by shalota1** [comment]
Hello I have the same problem any workaround for the moment with this issue ?

**Comment by shalota1** [comment]
> > Hello I have the same problem any workaround for the moment with this issue ?
> 
> Use Firefox to save the settings as outlined in my original bug report.

I solved it finally thanks :)

**Comment by elhananjair** [comment]
> Fixed it .... had a permission issue. (Copy of the ssl certificate / privkey wasn't owned by the turnserver user)
> 
> Note: chown -R turnserver:turnserver /etc/coturn/ chmod -R 755 /etc/coturn/ chmod 644 /etc/coturn/certs/$FqdnTurnServer.fullchain.pem chmod 600 /etc/coturn/certs/$FqdnTurnServer.privkey.pem
> 
> Source: https://blog.wydler.eu/2020/04/04/eigener-turn-server-fuer-jitsi-meet-bereitstellen
> 
> ICE candidate: OK!

Thank you @HeyWatchOutDude that is very helpful, I can now connect successfully with turn server. In my case though I am using Fedora Server and installed coturn package from fedora repo. I just changed user name specified in coturn.service which is coturn and changed it to root since the generated certificates are owned by root in my case.

**Comment by gohrner** [comment]
> I just changed user name specified in coturn.service which is coturn and changed it to root

@elhananjair: Running services as root user, especially services allowing external connections, as the root user is an extremely bad idea security wise.

Any security vulnerability in this service which allows an attacker to modify files will allow her/him to modify any file on your server, potentially providing the attacker full remote access (if it can be controlled which files shall be overwritten).

Any remote code execution vulnerability in this service will give any attacker full root permissions on your server immediately.

I'm pretty sure you won't want that.

The proper fix is to adjust / correct the certificate permissions, so the coturn server has at least read access to them while running under its own service account.

**Comment by elhananjair** [comment]
> > I just changed user name specified in coturn.service which is coturn and changed it to root
> 
> @elhananjair: Running services as root user, especially services allowing external connections, as the root user is an extremely bad idea security wise.
> 
> Any security vulnerability in this service which allows an attacker to modify files will allow her/him to modify any file on your server, potentially providing the attacker full remote access (if it can be controlled which files shall be overwritten).
> 
> Any remote code execution vulnerability in this service will give any attacker full root permissions on your server immediately.
> 
> I'm pretty sure you won't want that.
> 
> The proper fix is to adjust / correct the certificate permissions, so the coturn server has at least read access to them while running under its own service account.

Yeah you are right, I was just exhausted with the errors am facing, but I will change the user right now. Thank you so much

**Comment by HeyWatchOutDude** [comment]
> > Fixed it .... had a permission issue. (Copy of the ssl certificate / privkey wasn't owned by the turnserver user)
> > Note: chown -R turnserver:turnserver /etc/coturn/ chmod -R 755 /etc/coturn/ chmod 644 /etc/coturn/certs/$FqdnTurnServer.fullchain.pem chmod 600 /etc/coturn/certs/$FqdnTurnServer.privkey.pem
> > Source: https://blog.wydler.eu/2020/04/04/eigener-turn-server-fuer-jitsi-meet-bereitstellen
> > ICE candidate: OK!
> 
> Thank you @HeyWatchOutDude that is very helpful, I can now connect successfully with turn server. In my case though I am using Fedora Server and installed coturn package from fedora repo. I just changed user name specified in coturn.service which is coturn and changed it to root since the generated certificates are owned by root in my case.

Still working for you? Mine stopped ... NC version: 24.0.5 and Talk version: 14.0.5.

**Comment by elhananjair** [comment]
@HeyWatchOutDude yes it's working but still with the root user owning the certificate and running the service with root privilege. I have recently updated all sub-systems for HPB, Janus, nats-server, and signaling server. What problem are you facing?

**Comment by HeyWatchOutDude** [comment]
> @HeyWatchOutDude yes it's working but still with the root user owning the certificate and running the service with root privilege. I have recently updated all sub-systems for HPB, Janus, nats-server, and signaling server. What problem are you facing?

Strange :/ Im now getting again the error message "Error: No working ICE candidates returned by the TURN server."

**Comment by elhananjair** [comment]
> > @HeyWatchOutDude yes it's working but still with the root user owning the certificate and running the service with root privilege. I have recently updated all sub-systems for HPB, Janus, nats-server, and signaling server. What problem are you facing?
> 
> Strange :/ Im now getting again the error message "Error: No working ICE candidates returned by the TURN server."

If you are testing it on Firefox there might be still an issue reporting false results while talk works fine. If that's not the case, try to check a log from your TURN server.

**Comment by HeyWatchOutDude** [comment]
Error message when executing command "turnserver -a" (Note: "coturn" service is installed on the reverse proxy, worked before.)

Log:
`0: log file opened: /var/log/turn_11715_2022-10-02.log
0: 0 bytes per second allowed, combined server capacity
0: ERROR: 
CONFIG ERROR: Empty cli-password, and so telnet cli interface is disabled! Please set a non empty cli-password!
bind: Address already in use
Cannot bind local socket to addr: Address already in use
bind: Address already in use
Cannot bind local socket to addr: Address already in use
bind: Address already in use
Cannot bind local socket to addr: Address already in use
bind: Address already in use
Cannot bind local socket to addr: Address already in use
bind: Address already in use
bind: Address already in use
Cannot bind local socket to addr: Address already in use
Cannot bind local socket to addr: Address already in use
bind: Address already in use
Cannot bind local socket to addr: Address already in use
bind: Address already in use
Cannot bind local socket to addr: Address already in use
bind: Address already in use
bind: Address already in use
.....`

**Comment by gohrner** [comment]
> `bind: Address already in use`

Looks as if your service is still running and the instance you tried to start on the command line thus could't.

**Comment by HeyWatchOutDude** [comment]
> > `bind: Address already in use`
> 
> Looks as if your service is still running and the instance you tried to start on the command line thus could't.

I have changed the TURN Server URL from "cloud.example.com:5349" to "Internal IP of the Reverse Proxy:5349"  ... now im getting a working ICE candidate but sadly external calls are still not working.

**Comment by prismopensource** [comment]
Bug is still present in latest Nextcloud 25.0.5 with Talk 15.0.5 I'm afraid.

**Comment by prismopensource** [comment]
Since this bug is still present in 27.0.2 is there any chance to get this out of the backlog and into the 27.0.3 milestone?

**Comment by nickvergessen** [comment]
This is not in the backlog, but considered fixed, but we didn't backport it.
